libs/cardano-ledger-binary/testlib/Test/Cardano/Ledger/Binary/Cddl.hs

```diff
@@ -0,0 +1,238 @@
+cddlFailure diagCbor err =
+  expectationFailure $
+    unlines
+      [ "Failed to Cddl RoundTrip verification:"
+      , show err
+      , "Generated diag: " <> BSL8.unpack (unDiagCbor diagCbor)
+      ]
+
+cddlRoundTripAnnCborSpec ::
+  forall a.
+  (HasCallStack, Eq a, Show a, EncCBOR a, DecCBOR (Annotator a)) =>
+  -- | Cddl variable name
+  Version ->
+  T.Text ->
+  SpecWith CddlData
+cddlRoundTripAnnCborSpec version varName =
+  let lbl = label (Proxy @(Annotator a))
+   in it (T.unpack $ varName <> ": " <> lbl) $ \cddlData ->
+        withCddlVarFile varName cddlData $
+          cddlRoundTripAnnExpectation lbl version version (cborTrip @a)
+
+cddlRoundTripAnnExpectation ::
+  forall a.
+  (HasCallStack, Show a, Eq a) =>
+  T.Text ->
+  Version ->
+  Version ->
+  Trip a (Annotator a) ->
+  CddlVarFile ->
+  Expectation
+cddlRoundTripAnnExpectation lbl encVersion decVersion Trip {..} CddlVarFile {..} = do
+  forM_ cddlVarDiagCbor $ \diagCbor -> do
+    Cbor cbor <- diagToCbor diagCbor
+    let mkFailure encoding =
+          RoundTripFailure encVersion decVersion encoding cbor
+    case decodeFullAnnotator decVersion lbl tripDecoder cbor of
+      Left decErr ->
+        cddlFailure diagCbor $ mkFailure mempty Nothing Nothing Nothing (Just decErr)
+      Right val ->
+        let encoding = toPlainEncoding encVersion $ tripEncoder val
+         in case decodeAnnExtra lbl encVersion decVersion tripDecoder encoding of
+              Right (val', encodedBytes) ->
+                validateCddlConformance cddlVarFilePath encodedBytes >>= \case
+                  Left confErr ->
+                    cddlFailure diagCbor $
+                      mkFailure encoding (Just encodedBytes) (Just confErr) Nothing Nothing
+                  Right _bsl -> val' `shouldBe` val
+              Left embedErr -> cddlFailure diagCbor embedErr
+
+genCddlDiagCbor :: HasCallStack => Int -> Cddl -> IO [DiagCbor]
+genCddlDiagCbor numCases =
+  fmap (either error (map DiagCbor . BSL8.lines))
+    . readProcessNoFailure "generating examples" (proc "cddl" ["-", "generate", show numCases])
+    . unCddl
+
+diagToCbor :: HasCallStack => DiagCbor -> IO Cbor
+diagToCbor =
+  fmap (either error Cbor)
+    . readProcessNoFailure "converting cbor diagnostic notation to bytes" (proc "diag2cbor.rb" ["-"])
+    . unDiagCbor
+
+-- | Run a @cddl@ process and validate that encoded data is in confomrance with the
+-- supplied CDDL spec
+validateCddlConformance ::
+  -- | File path to the file with CDDL spec
+  FilePath ->
+  -- | CBOR encoded data that should conform with CDDL spec
+  BSL8.ByteString ->
+  IO (Either String BSL.ByteString)
+validateCddlConformance filePath =
+  readProcessNoFailure "validating cddl conformace" (proc "cddl" [filePath, "validate", "-"])
+
+-- | Run a process with `readProcess` and explode on failure. Accepts lazy bytestring as
+-- input for the spawned process. In case of a successfull process exit, but nonempty
+-- stderr, it will be printed out.
```

Comment:
Printed to where? And the bytestring in the successful return is what?

Reply:
In fact, this looks wrong, we are throwing away the stderr in the success case.

Reply:
> we are throwing away the stderr in the success case.

This is an unfortunate case, which is described in the comment. Using a cddl file results in a tons of warnings, since we are using only a small portion of the cddl. 

We might wanna find a solution where we only use a portion of the cddl that useful for a particular test case, but that is not possible without parsing the cddl file (or generating one from scratch)

So, for that reason this function does exactly the same as it did before. I just rewrote it using the `typed-process`, rather than inferior and less maintained `process-extras`


That being said, the haddock is indeed wrong, because it reflects my initial intention, rather than current behavior. I'll fix it.

Reply:
Done